File: bun_backend/server/services/runner.ts
import {
  killProcess,
  mergeSpawnEnv,
  spawnShell,
  type SpawnedProcess,
} from "../../native/process"
import { appsRepo } from "@db/apps"
import { configSetsRepo } from "@db/config-sets"
import { envVarsRepo } from "@db/env-vars"
import { runConfigsRepo } from "@db/run-configs"
import type { RunCommand, RunMode } from "@db/types"
import { matchReadyUrl } from "./ready-url"
import { applyTemplates, restoreTemplates } from "./templates"

export type ProcessStatus = "pending" | "running" | "exited" | "killed" | "error"

export type ProcessState = {
  commandId: number
  label: string
  command: string
  status: ProcessStatus
  exitCode: number | null
  pid: number | null
  /** URLs detected from process logs (Vite, Spring Boot, etc.) */
  urls: string[]
}

export type LogEvent = {
  type: "log"
  commandId: number
  stream: "stdout" | "stderr" | "system"
  text: string
  ts: number
}

export type StatusEvent = {
  type: "status"
  sessionId: string
  appId: number
  running: boolean
  processes: ProcessState[]
  error?: string
  ts: number
}

export type RunnerEvent = LogEvent | StatusEvent

type Session = {
  id: string
  appId: number
  mode: RunMode
  processes: ProcessState[]
  children: Map<number, SpawnedProcess>
  running: boolean
  abortSequential: boolean
  restored: boolean
  logBuffer: LogEvent[]
}

type RunnerStore = {
  sessions: Map<number, Session>
  listeners: Map<number, Set<(event: RunnerEvent) => void>>
}

const globalKey = "__workspaceManagerStore" as const

function getStore(): RunnerStore {
  const g = globalThis as typeof globalThis & {
    [globalKey]?: RunnerStore
  }
  if (!g[globalKey]) {
    g[globalKey] = {
      sessions: new Map(),
      listeners: new Map(),
    }
  }
  return g[globalKey]
}

const store = getStore()
const sessions = store.sessions
const globalListeners = store.listeners

const MAX_LOG_BUFFER = 5000

function emit(session: Session, event: RunnerEvent) {
  if (event.type === "log") {
    session.logBuffer.push(event)
    if (session.logBuffer.length > MAX_LOG_BUFFER) {
      session.logBuffer.splice(0, session.logBuffer.length - MAX_LOG_BUFFER)
    }
  }

  const globals = globalListeners.get(session.appId)
  if (!globals) return
  for (const listener of globals) {
    try {
      listener(event)
    } catch {
      // ignore broken subscribers
    }
  }
}

function statusEvent(session: Session, error?: string): StatusEvent {
  return {
    type: "status",
    sessionId: session.id,
    appId: session.appId,
    running: session.running,
    processes: session.processes.map((p) => ({ ...p })),
    error,
    ts: Date.now(),
  }
}

function systemLog(session: Session, commandId: number, text: string) {
  emit(session, {
    type: "log",
    commandId,
    stream: "system",
    text: text.endsWith("\n") ? text : `${text}\n`,
    ts: Date.now(),
  })
}

function noteReadyUrl(session: Session, commandId: number, line: string) {
  if (!session.running) return
  const match = matchReadyUrl(line)
  if (!match) return
  const processState = session.processes.find((p) => p.commandId === commandId)
  if (!processState) return
  if (processState.urls.includes(match.url)) return
  processState.urls.push(match.url)
  systemLog(
    session,
    commandId,
    `Detected URL (${match.label}): ${match.url}`
  )
  emit(session, statusEvent(session))
}

function clearReadyUrls(session: Session) {
  for (const processState of session.processes) {
    processState.urls = []
  }
}

function emitLogLine(
  session: Session,
  commandId: number,
  kind: "stdout" | "stderr",
  text: string
) {
  emit(session, {
    type: "log",
    commandId,
    stream: kind,
    text: text.endsWith("\n") ? text : `${text}\n`,
    ts: Date.now(),
  })
  noteReadyUrl(session, commandId, text)
}

async function pipeStream(
  session: Session,
  commandId: number,
  stream: ReadableStream<Uint8Array> | null,
  kind: "stdout" | "stderr"
) {
  if (!stream) return
  const reader = stream.getReader()
  const decoder = new TextDecoder()
  // Keep the raw buffer so partial ESC sequences are never stripped mid-chunk.
  let pending = ""
  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      const chunk = decoder.decode(value, { stream: true })
      if (!chunk) continue
      pending += chunk
      const parts = pending.split(/\r?\n/)
      pending = parts.pop() ?? ""
      for (const line of parts) {
        emitLogLine(session, commandId, kind, Bun.stripANSI(line))
      }
    }
    pending += decoder.decode()
    const rest = Bun.stripANSI(pending)
    if (rest) {
      emitLogLine(session, commandId, kind, rest)
    }
  } catch {
    // process killed / stream closed
  }
}

async function spawnCommand(
  session: Session,
  cmd: RunCommand,
  cwd: string,
  env: Record<string, string>
): Promise<number> {
  const processState = session.processes.find((p) => p.commandId === cmd.id)!
  processState.status = "running"
  emit(session, statusEvent(session))
  systemLog(session, cmd.id, `$ ${cmd.command}`)

  let child: SpawnedProcess
  try {
    child = spawnShell({ command: cmd.command, cwd, env })
  } catch (err) {
    processState.status = "error"
    processState.exitCode = 1
    systemLog(
      session,
      cmd.id,
      `Failed to start: ${err instanceof Error ? err.message : String(err)}`
    )
    emit(session, statusEvent(session))
    return 1
  }

  processState.pid = child.pid
  session.children.set(cmd.id, child)
  emit(session, statusEvent(session))

  const pipes = Promise.all([
    pipeStream(session, cmd.id, child.stdout, "stdout"),
    pipeStream(session, cmd.id, child.stderr, "stderr"),
  ])

  const exitCode = await child.exited
  await pipes
  session.children.delete(cmd.id)

  if (processState.status === "killed") {
    emit(session, statusEvent(session))
    return exitCode
  }

  processState.status = exitCode === 0 ? "exited" : "error"
  processState.exitCode = exitCode
  systemLog(
    session,
    cmd.id,
    exitCode === 0
      ? `Process exited with code ${exitCode}`
      : `Process failed with code ${exitCode}`
  )
  emit(session, statusEvent(session))
  return exitCode
}

async function runSession(session: Session) {
  const app = appsRepo.get(session.appId)
  if (!app) {
    session.running = false
    emit(session, statusEvent(session, "App not found"))
    return
  }

  try {
    applyTemplates(session.appId, session.id)
    systemLog(
      session,
      session.processes[0]?.commandId ?? 0,
      "Templates applied"
    )
  } catch (err) {
    session.running = false
    const message = err instanceof Error ? err.message : String(err)
    systemLog(
      session,
      session.processes[0]?.commandId ?? 0,
      `Template apply failed: ${message}`
    )
    emit(session, statusEvent(session, `Template apply failed: ${message}`))
    return
  }

  const set = configSetsRepo.resolveActive(session.appId)
  const env = mergeSpawnEnv(envVarsRepo.toRecord(set.id))
  const config = runConfigsRepo.getByConfigSet(set.id)
  const commands = config?.commands ?? []

  try {
    if (session.mode === "parallel") {
      await Promise.all(
        commands.map((cmd) => spawnCommand(session, cmd, app.project_path, env))
      )
    } else {
      for (const cmd of commands) {
        if (session.abortSequential) break
        const code = await spawnCommand(session, cmd, app.project_path, env)
        if (code !== 0) {
          systemLog(
            session,
            cmd.id,
            "Sequential run stopped due to non-zero exit"
          )
          break
        }
      }
    }
  } finally {
    session.running = false
    clearReadyUrls(session)
    if (!session.restored) {
      restoreTemplates(session.appId, session.id)
      session.restored = true
      systemLog(
        session,
        session.processes[0]?.commandId ?? 0,
        "Original files restored"
      )
    }
    emit(session, statusEvent(session))
  }
}

function createSession(appId: number): Session {
  const set = configSetsRepo.resolveActive(appId)
  const config = runConfigsRepo.getOrCreate(set.id)
  if (config.commands.length === 0) {
    throw new Error("No run commands configured")
  }

  return {
    id: `${appId}-${Date.now()}`,
    appId,
    mode: config.mode,
    processes: config.commands.map((cmd) => ({
      commandId: cmd.id,
      label: cmd.label || cmd.command,
      command: cmd.command,
      status: "pending",
      exitCode: null,
      pid: null,
      urls: [],
    })),
    children: new Map(),
    running: true,
    abortSequential: false,
    restored: false,
    logBuffer: [],
  }
}

export const runner = {
  getStatus(appId: number): StatusEvent | null {
    const session = sessions.get(appId)
    if (!session) return null
    return statusEvent(session)
  },

  getSnapshot(appId: number): {
    status: StatusEvent
    logs: LogEvent[]
  } {
    const session = sessions.get(appId)
    if (!session) {
      return {
        status: {
          type: "status",
          sessionId: "",
          appId,
          running: false,
          processes: [],
          ts: Date.now(),
        },
        logs: [],
      }
    }
    return {
      status: statusEvent(session),
      logs: [...session.logBuffer],
    }
  },

  subscribe(
    appId: number,
    listener: (event: RunnerEvent) => void,
    opts?: { replay?: boolean }
  ): () => void {
    let set = globalListeners.get(appId)
    if (!set) {
      set = new Set()
      globalListeners.set(appId, set)
    }
    set.add(listener)

    if (opts?.replay !== false) {
      const session = sessions.get(appId)
      if (session) {
        listener(statusEvent(session))
        for (const log of session.logBuffer) {
          listener(log)
        }
      } else {
        listener({
          type: "status",
          sessionId: "",
          appId,
          running: false,
          processes: [],
          ts: Date.now(),
        })
      }
    }

    return () => {
      set?.delete(listener)
      if (set && set.size === 0) globalListeners.delete(appId)
    }
  },

  async start(appId: number): Promise<StatusEvent> {
    if (!appsRepo.get(appId)) throw new Error("App not found")

    const existing = sessions.get(appId)
    if (existing?.running) {
      throw new Error("App is already running")
    }

    const session = createSession(appId)
    sessions.set(appId, session)
    emit(session, statusEvent(session))
    void runSession(session)
    return statusEvent(session)
  },

  async stop(appId: number): Promise<StatusEvent> {
    const session = sessions.get(appId)
    if (!session) throw new Error("No active session")

    session.abortSequential = true
    session.running = false
    clearReadyUrls(session)
    for (const processState of session.processes) {
      if (processState.status === "running") {
        processState.status = "killed"
      }
    }
    const children = [...session.children.values()]
    session.children.clear()
    // Emit cleared URLs immediately so UI updates before kill finishes
    emit(session, statusEvent(session))
    await Promise.all(children.map((child) => killProcess(child)))

    if (!session.restored) {
      restoreTemplates(appId, session.id)
      session.restored = true
      systemLog(
        session,
        session.processes[0]?.commandId ?? 0,
        "Stopped — original files restored"
      )
    }
    emit(session, statusEvent(session))
    return statusEvent(session)
  },

  async reload(appId: number): Promise<StatusEvent> {
    const existing = sessions.get(appId)
    if (existing?.running) {
      await this.stop(appId)
    }
    return this.start(appId)
  },
}

File: wails_backend/internal/services/runner.go
package services

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"os/exec"
	"regexp"
	"sync"
	"time"

	"wails_backend/internal/db"
	"wails_backend/internal/native"
	"wails_backend/internal/types"
)

const maxLogBuffer = 5000

type childProc struct {
	cmd *exec.Cmd
}

type session struct {
	id              string
	appID           int64
	mode            string
	processes       []types.ProcessState
	children        map[int64]*childProc
	running         bool
	abortSequential bool
	restored        bool
	logBuffer       []types.LogEvent
}

type Runner struct {
	db        *db.DB
	broadcast func(appID int64, event any)

	mu       sync.Mutex
	sessions map[int64]*session
}

func NewRunner(d *db.DB, broadcast func(appID int64, event any)) *Runner {
	return &Runner{db: d, broadcast: broadcast, sessions: map[int64]*session{}}
}

func (r *Runner) emit(s *session, event any) {
	if log, ok := event.(types.LogEvent); ok {
		s.logBuffer = append(s.logBuffer, log)
		if len(s.logBuffer) > maxLogBuffer {
			s.logBuffer = s.logBuffer[len(s.logBuffer)-maxLogBuffer:]
		}
	}
	if r.broadcast != nil {
		r.broadcast(s.appID, event)
	}
}

func (r *Runner) statusEvent(s *session, errMsg string) types.StatusEvent {
	procs := make([]types.ProcessState, len(s.processes))
	copy(procs, s.processes)
	for i := range procs {
		if procs[i].URLs == nil {
			procs[i].URLs = []string{}
		}
	}
	ev := types.StatusEvent{
		Type: "status", SessionID: s.id, AppID: s.appID,
		Running: s.running, Processes: procs, Ts: time.Now().UnixMilli(),
	}
	if errMsg != "" {
		ev.Error = errMsg
	}
	return ev
}

func (r *Runner) systemLog(s *session, commandID int64, text string) {
	if !stringsHasNL(text) {
		text += "\n"
	}
	r.emit(s, types.LogEvent{
		Type: "log", AppID: s.appID, CommandID: commandID, Stream: "system",
		Text: text, Ts: time.Now().UnixMilli(),
	})
}

func stringsHasNL(s string) bool {
	return len(s) > 0 && s[len(s)-1] == '\n'
}

func (r *Runner) noteReadyURL(s *session, commandID int64, line string) {
	if !s.running {
		return
	}
	match := MatchReadyURL(context.Background(), r.db, line)
	if match == nil {
		return
	}
	for i := range s.processes {
		if s.processes[i].CommandID != commandID {
			continue
		}
		for _, u := range s.processes[i].URLs {
			if u == match.URL {
				return
			}
		}
		s.processes[i].URLs = append(s.processes[i].URLs, match.URL)
		r.systemLog(s, commandID, fmt.Sprintf("Detected URL (%s): %s", match.Label, match.URL))
		r.emit(s, r.statusEvent(s, ""))
		return
	}
}

func (r *Runner) clearReadyURLs(s *session) {
	for i := range s.processes {
		s.processes[i].URLs = []string{}
	}
}

var ansiRe = regexp.MustCompile(`\x1b\[[0-9;?]*[A-Za-z]|\x1b\][^\x07]*\x07`)

func stripANSI(s string) string {
	return ansiRe.ReplaceAllString(s, "")
}

func (r *Runner) emitLogLine(s *session, commandID int64, kind, text string) {
	if !stringsHasNL(text) {
		text += "\n"
	}
	r.emit(s, types.LogEvent{
		Type: "log", AppID: s.appID, CommandID: commandID, Stream: kind,
		Text: text, Ts: time.Now().UnixMilli(),
	})
	r.noteReadyURL(s, commandID, text)
}

func (r *Runner) pipeStream(s *session, commandID int64, reader io.Reader, kind string) {
	buf := bufio.NewReader(reader)
	var pending string
	for {
		chunk, err := buf.ReadString('\n')
		pending += chunk
		for {
			i := indexNL(pending)
			if i < 0 {
				break
			}
			line := stripANSI(trimCR(pending[:i]))
			pending = pending[i+1:]
			r.emitLogLine(s, commandID, kind, line)
		}
		if err != nil {
			rest := stripANSI(trimCR(pending))
			if rest != "" {
				r.emitLogLine(s, commandID, kind, rest)
			}
			return
		}
	}
}

func indexNL(s string) int {
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			return i
		}
	}
	return -1
}

func trimCR(s string) string {
	if len(s) > 0 && s[len(s)-1] == '\r' {
		return s[:len(s)-1]
	}
	return s
}

func (r *Runner) spawnCommand(s *session, cmd types.RunCommand, cwd string, env []string) int {
	var processState *types.ProcessState
	for i := range s.processes {
		if s.processes[i].CommandID == cmd.ID {
			processState = &s.processes[i]
			break
		}
	}
	if processState == nil {
		return 1
	}
	processState.Status = "running"
	r.emit(s, r.statusEvent(s, ""))
	r.systemLog(s, cmd.ID, "$ "+cmd.Command)

	child, err := native.SpawnShell(cmd.Command, cwd, env)
	if err != nil {
		processState.Status = "error"
		code := int64(1)
		processState.ExitCode = &code
		r.systemLog(s, cmd.ID, "Failed to start: "+err.Error())
		r.emit(s, r.statusEvent(s, ""))
		return 1
	}
	stdout, err := child.StdoutPipe()
	if err != nil {
		processState.Status = "error"
		code := int64(1)
		processState.ExitCode = &code
		r.systemLog(s, cmd.ID, "Failed to start: "+err.Error())
		r.emit(s, r.statusEvent(s, ""))
		return 1
	}
	stderr, err := child.StderrPipe()
	if err != nil {
		processState.Status = "error"
		code := int64(1)
		processState.ExitCode = &code
		r.systemLog(s, cmd.ID, "Failed to start: "+err.Error())
		r.emit(s, r.statusEvent(s, ""))
		return 1
	}
	if err := child.Start(); err != nil {
		processState.Status = "error"
		code := int64(1)
		processState.ExitCode = &code
		r.systemLog(s, cmd.ID, "Failed to start: "+err.Error())
		r.emit(s, r.statusEvent(s, ""))
		return 1
	}
	pid := int64(child.Process.Pid)
	processState.PID = &pid
	r.mu.Lock()
	s.children[cmd.ID] = &childProc{cmd: child}
	r.mu.Unlock()
	r.emit(s, r.statusEvent(s, ""))

	var wg sync.WaitGroup
	wg.Add(2)
	go func() { defer wg.Done(); r.pipeStream(s, cmd.ID, stdout, "stdout") }()
	go func() { defer wg.Done(); r.pipeStream(s, cmd.ID, stderr, "stderr") }()
	waitErr := child.Wait()
	wg.Wait()

	r.mu.Lock()
	delete(s.children, cmd.ID)
	r.mu.Unlock()

	if processState.Status == "killed" {
		r.emit(s, r.statusEvent(s, ""))
		return 1
	}
	exitCode := 0
	if waitErr != nil {
		if ee, ok := waitErr.(*exec.ExitError); ok {
			exitCode = ee.ExitCode()
		} else {
			exitCode = 1
		}
	}
	code64 := int64(exitCode)
	processState.ExitCode = &code64
	if exitCode == 0 {
		processState.Status = "exited"
		r.systemLog(s, cmd.ID, fmt.Sprintf("Process exited with code %d", exitCode))
	} else {
		processState.Status = "error"
		r.systemLog(s, cmd.ID, fmt.Sprintf("Process failed with code %d", exitCode))
	}
	r.emit(s, r.statusEvent(s, ""))
	return exitCode
}

func (r *Runner) runSession(s *session) {
	ctx := context.Background()
	app, err := r.db.GetAppT(ctx, s.appID)
	if err != nil {
		s.running = false
		r.emit(s, r.statusEvent(s, "App not found"))
		return
	}
	if err := ApplyTemplates(ctx, r.db, s.appID, s.id); err != nil {
		s.running = false
		msg := err.Error()
		cmdID := int64(0)
		if len(s.processes) > 0 {
			cmdID = s.processes[0].CommandID
		}
		r.systemLog(s, cmdID, "Template apply failed: "+msg)
		r.emit(s, r.statusEvent(s, "Template apply failed: "+msg))
		return
	}
	cmdID := int64(0)
	if len(s.processes) > 0 {
		cmdID = s.processes[0].CommandID
	}
	r.systemLog(s, cmdID, "Templates applied")

	set, err := r.db.ResolveActive(ctx, s.appID)
	if err != nil {
		s.running = false
		r.emit(s, r.statusEvent(s, err.Error()))
		return
	}
	envMap, _ := r.db.EnvToRecord(ctx, set.ID)
	env := native.MergeSpawnEnv(envMap)
	config, err := r.db.GetRunConfigByConfigSetT(ctx, set.ID)
	if err != nil {
		s.running = false
		r.emit(s, r.statusEvent(s, err.Error()))
		return
	}
	var commands []types.RunCommand
	if config != nil {
		commands = config.Commands
	}

	func() {
		defer func() {
			s.running = false
			r.clearReadyURLs(s)
			if !s.restored {
				_ = RestoreTemplates(ctx, r.db, s.appID, s.id)
				s.restored = true
				r.systemLog(s, cmdID, "Original files restored")
			}
			r.emit(s, r.statusEvent(s, ""))
		}()
		if s.mode == "parallel" {
			var wg sync.WaitGroup
			for _, c := range commands {
				c := c
				wg.Add(1)
				go func() { defer wg.Done(); r.spawnCommand(s, c, app.ProjectPath, env) }()
			}
			wg.Wait()
			return
		}
		for _, c := range commands {
			if s.abortSequential {
				break
			}
			code := r.spawnCommand(s, c, app.ProjectPath, env)
			if code != 0 {
				r.systemLog(s, c.ID, "Sequential run stopped due to non-zero exit")
				break
			}
		}
	}()
}

func (r *Runner) createSession(ctx context.Context, appID int64) (*session, error) {
	set, err := r.db.ResolveActive(ctx, appID)
	if err != nil {
		return nil, err
	}
	config, err := r.db.GetOrCreateRunConfig(ctx, set.ID)
	if err != nil {
		return nil, err
	}
	if len(config.Commands) == 0 {
		return nil, fmt.Errorf("No run commands configured")
	}
	procs := make([]types.ProcessState, 0, len(config.Commands))
	for _, cmd := range config.Commands {
		label := cmd.Command
		if cmd.Label != nil && *cmd.Label != "" {
			label = *cmd.Label
		}
		procs = append(procs, types.ProcessState{
			CommandID: cmd.ID, Label: label, Command: cmd.Command,
			Status: "pending", URLs: []string{},
		})
	}
	return &session{
		id:        fmt.Sprintf("%d-%d", appID, time.Now().UnixMilli()),
		appID:     appID,
		mode:      config.Mode,
		processes: procs,
		children:  map[int64]*childProc{},
		running:   true,
	}, nil
}

func (r *Runner) GetStatus(appID int64) types.StatusEvent {
	r.mu.Lock()
	defer r.mu.Unlock()
	s := r.sessions[appID]
	if s == nil {
		return types.StatusEvent{Type: "status", AppID: appID, Processes: []types.ProcessState{}, Ts: time.Now().UnixMilli()}
	}
	return r.statusEvent(s, "")
}

func (r *Runner) GetSnapshot(appID int64) types.RunnerLogsSnapshot {
	r.mu.Lock()
	defer r.mu.Unlock()
	s := r.sessions[appID]
	if s == nil {
		return types.RunnerLogsSnapshot{
			Status: types.StatusEvent{Type: "status", AppID: appID, Processes: []types.ProcessState{}, Ts: time.Now().UnixMilli()},
			Logs:   []types.LogEvent{},
		}
	}
	logs := make([]types.LogEvent, len(s.logBuffer))
	copy(logs, s.logBuffer)
	return types.RunnerLogsSnapshot{Status: r.statusEvent(s, ""), Logs: logs}
}

func (r *Runner) Start(ctx context.Context, appID int64) (types.StatusEvent, error) {
	if _, err := r.db.GetAppT(ctx, appID); err != nil {
		return types.StatusEvent{}, err
	}
	r.mu.Lock()
	existing := r.sessions[appID]
	if existing != nil && existing.running {
		r.mu.Unlock()
		return types.StatusEvent{}, fmt.Errorf("App is already running")
	}
	r.mu.Unlock()

	s, err := r.createSession(ctx, appID)
	if err != nil {
		return types.StatusEvent{}, err
	}
	r.mu.Lock()
	r.sessions[appID] = s
	r.mu.Unlock()
	r.emit(s, r.statusEvent(s, ""))
	go r.runSession(s)
	return r.statusEvent(s, ""), nil
}

func (r *Runner) Stop(ctx context.Context, appID int64) (types.StatusEvent, error) {
	r.mu.Lock()
	s := r.sessions[appID]
	if s == nil {
		r.mu.Unlock()
		return types.StatusEvent{}, fmt.Errorf("No active session")
	}
	s.abortSequential = true
	s.running = false
	r.clearReadyURLs(s)
	for i := range s.processes {
		if s.processes[i].Status == "running" {
			s.processes[i].Status = "killed"
		}
	}
	children := make([]*childProc, 0, len(s.children))
	for _, c := range s.children {
		children = append(children, c)
	}
	s.children = map[int64]*childProc{}
	r.mu.Unlock()

	r.emit(s, r.statusEvent(s, ""))
	for _, c := range children {
		if c.cmd != nil && c.cmd.Process != nil {
			_ = native.KillPid(c.cmd.Process.Pid)
		}
	}
	if !s.restored {
		_ = RestoreTemplates(ctx, r.db, appID, s.id)
		s.restored = true
		cmdID := int64(0)
		if len(s.processes) > 0 {
			cmdID = s.processes[0].CommandID
		}
		r.systemLog(s, cmdID, "Stopped — original files restored")
	}
	r.emit(s, r.statusEvent(s, ""))
	return r.statusEvent(s, ""), nil
}

func (r *Runner) Reload(ctx context.Context, appID int64) (types.StatusEvent, error) {
	r.mu.Lock()
	existing := r.sessions[appID]
	running := existing != nil && existing.running
	r.mu.Unlock()
	if running {
		if _, err := r.Stop(ctx, appID); err != nil {
			return types.StatusEvent{}, err
		}
	}
	return r.Start(ctx, appID)
}

func (r *Runner) StopAll(ctx context.Context) {
	r.mu.Lock()
	ids := make([]int64, 0, len(r.sessions))
	for id, s := range r.sessions {
		if s.running {
			ids = append(ids, id)
		}
	}
	r.mu.Unlock()
	for _, id := range ids {
		_, _ = r.Stop(ctx, id)
	}
}
